### PR TITLE
I don't know how to name this pull request now

### PR DESCRIPTION
**Pull request by Ascynx**
Adds a very cool discord rich presence

- Options
  - allows to disable the rpc
  - Ability to change the 1st line in the rpc to pretty much anything. (updates every 15 seconds)
  details1
  details2
  has multiple dynamic values that can be put inside the options:
   - for the player name
   - the shard
   - the name of the item held in main hand
   - the class used by the player
   - the location (if exists) of the player
  - The small image has a version for every shard "type" known currently in the game
	shards
- Cool factor + 1000

if needed I can provide access to the discord.com/developers application

### CODE REVIEW

**Comment by Ascynx** [comment]
yeah there's 3 others btw, was not completely sure if I was to open multiple pull requests (1 for every version branch) or not

**Comment by Njol** [comment]
Looks cool!
BTW if you think any additional data would be nice but is not currently available to the client, that could be added too.
As for the different versions: Unless you had to adapt code for them, a single PR on 1.16.4 is best as I merge changes from one branch to all others myself for every update anyway.

**Comment by Ascynx** [comment]
> 

I'll see if there's any need to adapt the code in between version but after the original pull request I noticed that at the current time it didn't require a change in code to make it work between 1.16.4 and 1.16.5, I'll check for 1.17 and 1.18 later today.
also closed the other 3 pull requests.
also also in the last commit a new page was added to get the locations from, the page is: https://gist.github.com/Ascynx/4b1602b907da62a9cca6f135fd334737 (html page) and contains currently every locations I deemed important in the format
```
SHARD: [
  "(+X:-Z)(-X:+Z)/name"
]
```

it's possible on user side to disable updates from that page from the locationUpdate setting, and if there's an error it's also possible for them to place the file themselves (if the locationUpdate setting is disabled it won't check if the commit is the same as the last one posted)

**Comment by Ascynx** [comment]
I think those are all the changes I can think of right now.
I also don't think there's anything that will break when moving in between versions.

**Comment by Njol** [comment]
I'll merge this once I finish my current feature and then make a new release :)

**Comment by Ascynx** [comment]
I'm currently also working on stuff but since it's not really related to the discord rpc I decided to move to a new [branch](https://github.com/Ascynx/UnofficialMonumentaMod/tree/quality-of-life), will also have a "changelog" file, which contains more informations on the changes.

**Comment by Ascynx** [comment]
Found a few bugs and till I push the fix commit, I'll keep it as draft

**Comment by Ascynx** [comment]
didn't think it would take this little time.

and it seems I forgot to remove a comment.

File: src/main/java/ch/njol/unofficialmonumentamod/mixins/ThreadExecutorMixin.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+
+@Mixin(ThreadExecutor.class)
+public abstract class ThreadExecutorMixin<R extends Runnable> {
+    /**
+     *  Silences errors that happen when a player leaves monumenta or changes shard.
+     */
+    @Inject(method = "executeTask", at = @At(value = "INVOKE", target = "Lorg/apache/logging/log4j/Logger;fatal(Ljava/lang/String;Ljava/lang/Object;Ljava/lang/Object;)V", remap = false), locals = LocalCapture.CAPTURE_FAILSOFT, cancellable = true)
+    private void silenceTeamFatal(R task, CallbackInfo ci, Exception exception) {
+        if (UnofficialMonumentaModClient.options.silenceTeamError && (exception.getMessage().matches("Player is either on another team or not on any team\\. Cannot remove from team 'players'\\.") || exception.getMessage().matches("Cannot invoke \"net\\.minecraft\\.scoreboard\\.Team\\.getName\\(\\)\" because \"team\" is null"))) {
```
Comment (Njol) [inline src/main/java/ch/njol/unofficialmonumentamod/mixins/ThreadExecutorMixin.java]: Why use regex for a simple text check? Just use a normal equals.

File: src/main/java/ch/njol/unofficialmonumentamod/options/ConfigMenu.java [hunk 2]
```diff
@@ -55,78 +59,87 @@ protected void init() {
...
+							.setSaveConsumer(saveConsumer::accept).build();
 					} else if (field.getType() == Options.DescriptionLine.class) {
-						category.addEntry(config.entryBuilder()
-							.startTextDescription(new TranslatableText(translateKey))
+						entry = builder
+							.startTextDescription(new TranslatableText(translateKey)).build();
 //                                .setTooltip(new TranslatableText(translateKey + ".tooltip"))
-							.build());
-					} else {
+					} else if (field.getType() == KeybindingHandler.Keybinding.class) {
```
Comment (Njol) [inline src/main/java/ch/njol/unofficialmonumentamod/options/ConfigMenu.java]: Is there a reason why you added key code options? Those can just be put into the normal Minecraft controls menu by making a new key binding.

File: src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java [hunk 3]
```diff
@@ -51,27 +78,68 @@ public void onInitializeClient() {
...
+			KeybindingHandler.tick();
+			Calculator.tick();
+			CooldownManager.update();
+		});
+
+		ClientTickEvents.END_WORLD_TICK.register(world -> {
+			Notifier.tick();
 		});
 
+		ClientCommandManager.DISPATCHER.register((ClientCommandManager.literal("spoof").then(ClientCommandManager.argument("name", StringArgumentType.greedyString()).executes(ItemNameSpoofer::commandNameSpoofer))));
```
Comment (Njol) [inline src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java]: I don't get what item name spoofing is supposed to accomplish? Custom item textures?

File: src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java [hunk 4]
```diff
@@ -51,27 +78,68 @@ public void onInitializeClient() {
...
 
 	}
 
+	public static boolean isOnMonumenta() {
+		if (onMonumenta != null) return onMonumenta;
+		Boolean onMM = null;
+
+		MinecraftClient mc = MinecraftClient.getInstance();
+		if (Locations.getShard() != null) onMM = true;
+		if (onMM == null) onMM = !mc.isInSingleplayer() && Objects.requireNonNull(mc.getCurrentServerEntry()).address.toLowerCase().matches("(?i)server.playmonumenta.com|monumenta-11.playmonumenta.com|monumenta-8.playmonumenta.com|monumenta-13.playmonumenta.com");
```
Comment (Njol) [inline src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java]: Should just check for endsWith(".playmonumenta.com") so any future subdomains work too (as long as the domain doesn't change)

File: src/main/java/ch/njol/unofficialmonumentamod/Utils.java [hunk 5]
```diff
@@ -17,10 +55,119 @@ public static String getPlainDisplayName(ItemStack itemStack) {
...
+        return UnofficialMonumentaModClient.options.chestsortDisabledEverywhereElse;
+    }
+
     public static float smoothStep(float f) {
         if (f <= 0) return 0;
         if (f >= 1) return 1;
         return f * f * (3 - 2 * f);
     }
 
+    public static class abilitiesDisplay {
```
Comment (Njol) [inline src/main/java/ch/njol/unofficialmonumentamod/Utils.java]: Why move this here? it makes no sense to be here.
Also preferably don't move my code around, it will lead to massive conflicts with the changes I'm currently working on

File: src/main/java/ch/njol/unofficialmonumentamod/misc/Locations.java [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+public class Locations {
+    //add shards with locations here
+    public ArrayList<String> VALLEY;
+    public ArrayList<String> PLOTS;
+    public ArrayList<String> ISLES;
+
+    private static final String CACHE_FILE_PATH = "monumenta/unofficial-monumenta-mod-locations.json";
+    private String update_commit;
+    private static final String UPDATE_GIST_URL = "https://api.github.com/gists/4b1602b907da62a9cca6f135fd334737";//put new locations in that gist
```
Comment (Njol) [inline src/main/java/ch/njol/unofficialmonumentamod/misc/Locations.java]: I dislike the mod downloading data. Either put it into the mod itself, or adapt the server to send it.

File: src/main/java/ch/njol/unofficialmonumentamod/misc/Locations.java [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+            String shard = null;
+
+            for (Text text : header.getSiblings()) {
+                if (text.getString().matches("<.*>")) {
+                    //player shard
+                    shard = text.getString().substring(1, text.getString().length() - 1);
+                }
+            }
+            return shard;
+        }catch (NullPointerException ignored) {}
```
Comment (Njol) [inline src/main/java/ch/njol/unofficialmonumentamod/misc/Locations.java]: wtf. Sorry, but please fix your NPEs instead of catching them lol

File: src/main/java/ch/njol/unofficialmonumentamod/misc/managers/KeybindingHandler.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
+package ch.njol.unofficialmonumentamod.misc.managers;
+
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.util.InputUtil;
+import net.minecraft.text.TranslatableText;
+import net.minecraft.text.Text;
+
+import java.util.ArrayList;
+
+public class KeybindingHandler {
```
Comment (Njol) [inline src/main/java/ch/njol/unofficialmonumentamod/misc/managers/KeybindingHandler.java]: Why is this needed? Minecraft already offers key bindings.

**Review by Njol — COMMENTED** [review]
Damn this PR has grown big.

I looked at it a bit and have some comments:
* Please don't change the indentation to spaces, keep the tabs.
* Doing tesseract cooldowns client-side won't work well due to how they work server-side. It would be better to modify the server to send some data when they're used, which will prevent having to update the client mod with every small change to tesseracts and would also prevent any false activations (as right now it seems to start the cooldown on any click with a tesseract, ignoring if it actually worked or not)

File: src/main/java/ch/njol/unofficialmonumentamod/options/ConfigMenu.java (hunk 2 above)
Comment (Ascynx) [reply src/main/java/ch/njol/unofficialmonumentamod/options/ConfigMenu.java]: at one point I wanted to do some things with keybinds but forgot to remove

File: src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java (hunk 3 above)
Comment (Ascynx) [reply src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java]: not really, just cosmetic.

File: src/main/java/ch/njol/unofficialmonumentamod/misc/managers/KeybindingHandler.java (hunk 8 above)
Comment (Ascynx) [reply src/main/java/ch/njol/unofficialmonumentamod/misc/managers/KeybindingHandler.java]: was linked directly to option thing, seemed like asking users to check 2 places for settings would be too much, but not like I'm gonna keep it anyway.

File: src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java (hunk 4 above)
Comment (Ascynx) [reply src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java]: that's just me being dumb

File: src/main/java/ch/njol/unofficialmonumentamod/Utils.java (hunk 5 above)
Comment (Ascynx) [reply src/main/java/ch/njol/unofficialmonumentamod/Utils.java]: I'm not completely sure why I did so, but I think I was reusing the methods elsewhere and ended up just removing the other places that called them.

File: src/main/java/ch/njol/unofficialmonumentamod/misc/Locations.java (hunk 7 above)
Comment (Ascynx) [reply src/main/java/ch/njol/unofficialmonumentamod/misc/Locations.java]: what's the dumbest is that there isn't even a reason to have that try catch block there.

File: src/main/java/ch/njol/unofficialmonumentamod/misc/Locations.java (hunk 6 above)
Comment (Ascynx) [reply src/main/java/ch/njol/unofficialmonumentamod/misc/Locations.java]: understandable, I just don't particularly like hardcoding values.

**Comment by Ascynx** [comment]
>     * Doing tesseract cooldowns client-side won't work well due to how they work server-side. It would be better to modify the server to send some data when they're used, which will prevent having to update the client mod with every small change to tesseracts and would also prevent any false activations (as right now it seems to start the cooldown on any click with a tesseract, ignoring if it actually worked or not)

I actually made it so it checks the tooltip of the item to see if it should or not trigger, at first It didn't check and just added the cooldown but here for most tesseracts it works like it should. (and by most I say pretty much all of them except the tesseract of light)

and the indent replacement was an accident while using the ide

**Comment by Ascynx** [comment]
I don't truly remember where the methods I moved into Utils.abilitiesDisplay are from so I'll fix that later

**Comment by Ascynx** [comment]
I don't think it's gonna stop a merge conflict but it's back where it was before.

File: src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java (hunk 3 above)
Comment (Ascynx) [reply src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java]: the only thing I can say about that is that it can look cool and I remember using a similar feature on another server to know when I lent my items if they were actually mine.

Comment (Njol) [reply src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java]: Since Monumenta items are not distinguishable from one another you won't be able to accomplish this. The best way is to infuse or stat track an item to mark it with your name, which doesn't require a client mod. I'm very hesitant to add this feature, as it's hardly useful and it'll probably fall onto me to keep supporting it.

File: src/main/java/ch/njol/unofficialmonumentamod/misc/managers/KeybindingHandler.java (hunk 8 above)
Comment (Njol) [reply src/main/java/ch/njol/unofficialmonumentamod/misc/managers/KeybindingHandler.java]: Almost all other mods also have keybindings there- I have exactly one mod that has keybindings in its settings, and that is because it offers better customisations with modifier keys.

**Comment by Njol** [comment]
The code looks much cleaner already, nice :)
I'll think a bit if all these features really make sense in the mod, mostly because I don't want to end up having to support them alone.

**Comment by Ascynx** [comment]
> The code looks much cleaner already, nice :) I'll think a bit if all these features really make sense in the mod, mostly because I don't want to end up having to support them alone.

I wouldn't have created a pull request if I would let someone else deal with the crashes.

and I just thought about how I implemented keybinds at one point and it would've just been easier to store as just the keycode add as actual keybinds and mirror into the settings.

and... for stuff with packets, I've never really worked with those and don't really know how to work with plugins so I preferred working with what I had access to.

File: src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java (hunk 3 above)
Comment (Ascynx) [reply src/main/java/ch/njol/unofficialmonumentamod/UnofficialMonumentaModClient.java]: It seems I didn't think that through

**Comment by Njol** [comment]
Whoops, completely forgot about this. Will merge soonish and maybe tweak it a bit idk. I'll also poke the RP team to finally get me the icons they promised over a month ago.

**Comment by Ascynx** [comment]
I didn't forget about it, but I might've moved branch

**Comment by Ascynx** [comment]
the one I worked on these past few days is https://github.com/Ascynx/UnofficialMonumentaMod/tree/discordrpc-1.18, I truly don't know if this branch works since the 1.18 update, which either I'll check or retroactively edit what I did here with the 1.18 version's changes.

EDIT: yeah it doesn't work on 1.18.2 (mixins/mapping issue)

**Comment by Njol** [comment]
I meant that I forgot about this - I was very busy with r3 stuff recently.
I will try to merge this PR this weekend. If you made any new features since you opened this PR, maybe include them now, or make a new PR later- I'll try to make releases more often now so it won't matter too much :)

**Comment by Ascynx** [comment]
ok

**Comment by Ascynx** [comment]
btw do I update to 1.18 with the discordrpc-1.18 branch's changes?

**Comment by Njol** [comment]
Update to whatever version of your code you deem is the one you want to see in an official release :)

I can probably handle any possible versioning issues, though would of course appreciate them be fixed.

**Comment by Ascynx** [comment]
> Update to whatever version of your code you deem is the one you want to see in an official release :)

The question was just since monumenta updated to 1.18 (I think 2 months ago?),
there doesn't seem to be a reason to maintain a version that can't be used on the server (mainly due to mixin/mapping issues).

**Comment by Ascynx** [comment]
Placed here for people that might be interested

as I said in my last commit, here's the documentation on how to override the shards (along with location config documentation)
https://github.com/Ascynx/UnofficialMonumentaMod/wiki

**Comment by Njol** [comment]
You're been hard at work with this, nice :)
Please tell me when it's ready. No need to rush, the RP team forgot the icons (again). They started doing something after I reminded them though, so _maybe_ they'll manage to throw something together today or maybe not. So the mod release may be delayed till wednesday as that's the earliest i got lots of free time again ^^
Unless of course you need more time still, then it can be delayed further.

**Comment by Ascynx** [comment]
> Please tell me when it's ready.

In a way it's ready in another it isn't, with time more ideas / bugs come. I'll just focus on bugfixes for now.
It will never end if I just implement every idea that comes through my mind.

**Comment by Ascynx** [comment]
setting chestplate as elytra doesn't 100% work (will still render cape) but it's funny so I'm keeping it that way.
<details>(the cape not being the normal one is just an asset edit)</details>

**Comment by Ascynx** [comment]
I currently have not encountered any major bugs so it should be fine.

By minor I mean rare rendering issues.

**Comment by Njol** [comment]
I'll merge this this Saturday, then merge my own changes, then tweak the code a bit after testing. Also, the RP team finally got their icons almost ready, so hopefully there'll be a release this weekend :)

**Review by Njol — APPROVED** [review]
